mono/mini/builtin-types.cs

```diff
@@ -0,0 +1,1371 @@
+		var y = (nfloat)20f;
+		var z = x + y;
+		if ((float)z != 30f)
+			return 1;
+		return 0;
+	}
+
+	static int test_0_nfloat_sub ()
+	{
+		var x = (nfloat)10f;
+		var y = (nfloat)20f;
+		var z = x - y;
+		if ((float)z != -10f)
+			return 1;
+		return 0;
+	}
+
+	static int test_0_nfloat_mul ()
+	{
+		var x = (nfloat)10f;
+		var y = (nfloat)20f;
+		var z = x * y;
+		if ((float)z != 200f)
+			return 1;
+		return 0;
+	}
+
+	static int test_0_nfloat_div ()
+	{
+		var x = (nfloat)30f;
+		var y = (nfloat)3f;
+		var z = x / y;
+		if ((float)z != 10f)
+			return 1;
+		return 0;
+	}
+
+	static int test_0_nfloat_rem ()
+	{
+		var x = (nfloat)22f;
+		var y = (nfloat)10f;
+		var z = x % y;
+		if ((float)z != 2f)
+			return 1;
+		return 0;
+	}
+
+	static int test_0_nfloat_cmp_same_val ()
+	{
+		var x = (nfloat)10f;
+		var y = (nfloat)10f;
+		if (!(x == y))
+			return 1;
+		if (x != y)
+			return 2;
+		if (x < y)
+			return 3;
+		if (x > y)
+			return 4;
+		if (!(x <= y))
+			return 5;
+		if (!(x >= y))
+			return 6;
+		return 0;
+	}
+
+	static int test_0_nfloat_cmp_small_val ()
+	{
+		var x = (nfloat)5f;
+		var y = (nfloat)10f;
+		if (x == y)
+			return 1;
+		if (!(x != y))
+			return 2;
+		if (!(x < y))
+			return 3;
+		if (x > y)
+			return 4;
+		if (!(x <= y))
+			return 5;
+		if (x >= y)
+			return 6;
+		return 0;
+	}
+
+	static int test_0_nfloat_cmp_large_val ()
+	{
+		var x = (nfloat)20f;
+		var y = (nfloat)10f;
+		if (x == y)
+			return 1;
+		if (!(x != y))
+			return 2;
+		if (x < y)
+			return 3;
+		if (!(x > y))
+			return 4;
+		if (x <= y)
+			return 1;
+		if (!(x >= y))
+			return 1;
+		return 0;
+	}
+
+	static int test_0_nfloat_cmp_left_nan ()
+	{
+		var x = (nfloat)float.NaN;
+		var y = (nfloat)10f;
+		if (x == y)
+			return 1;
+		if (!(x != y))
+			return 2;
+		if (x < y)
+			return 3;
+		if (x > y)
+			return 4;
+		if (x <= y)
+			return 1;
+		if (x >= y)
+			return 1;
+		return 0;
+	}
+
+
+	static int test_0_nfloat_cmp_right_nan ()
+	{
+		var x = (nfloat)10f;
+		var y = (nfloat)float.NaN;
+		if (x == y)
+			return 1;
+		if (!(x != y))
+			return 2;
+		if (x < y)
+			return 3;
+		if (x > y)
+			return 4;
+		if (x <= y)
+			return 1;
+		if (x >= y)
+			return 1;
+		return 0;
+	}
+	// static int test_0_nfloat_call_boxed_equals ()
+	// {
+	// 	object x = new nfloat (10f);
+	// 	object y = new nfloat (10f);
+	// 	if (!x.Equals (y))
+	// 		return 1;
+	// 	return 0;
+	// }
+
+	static int test_0_nfloat_call_boxed_funs ()
+	{
+		object x = new nfloat (10f);
+		object y = new nfloat (10f);
+		if (x.GetHashCode () == 0)
+			return 2;
+		if (x.ToString () != "10")
+			return 3;
+		return 0;
+	}
+
+	public int test_0_nfloat_unboxed_member_calls ()
+	{
+		var x = (nfloat)10f;
+		if (!x.Equals (x))
+			return 1;
+		if (x != nfloat.Parse ("10"))
+			return 2;
+		return 0;
+	}
+
+	public static int Main (String[] args) {
+		return TestDriver.RunTests (typeof (BuiltinTests), args);
+	}
+}
+
+
+// !!! WARNING - GENERATED CODE - DO NOT EDIT !!!
```

Comment:
I'm not sure what is worse: (1) not testing at all or (2) testing something potentially slightly off.

What prevents us to move this code from XI to mono nowadays? Except for the magic check, nothing else seems to be specific to the product.